Add "or" helper to output content if/or

JemBijoux opened the pull request:
While working with templates, it's often helpful to be able to output
content if any one of several arguments is valid. (For instance, only
show a product tab if there is warranty info OR shipping info. Only
show social icons wrapper if any one of the icons is present, etc.)

- Add or helper based off the existing "if" helper - content will
  render if any/at least one of the arguments supplied validates
- Add some basic tests to verify the functionality of the or helper

mcampa commented:
I think this should be done inside the `if` helper, to be used like 
```js
{{#if var1 'or' var2}} ... {{/if}}
```

take a look at https://github.com/bigcommerce/paper/blob/master/helpers/if.js#L42

JemBijoux commented:
how would you go about dealing with multiple arguments though? Maybe it's a matter or renaming to something like "any"? It gets trickier when you have say... five items you want to test against. (Use case being: say you have a tab on a product page with width, height, length, width, shipping details, and you want to only show the block if any or all of them are available?)

mcampa commented:
Makes sense. 

I would prefer if you call it "any", @hegrec what do you think? We talked about naming it "some" but I think "any" makes more sense

hegrec commented:
We also have this helper `#any` which does a similar thing except that it only looks into a single array passed in. We could either overload `#any` to accept multiple parameters and test each one if more than one argument is passed, or add this new one which does essentially `_.some`. Maybe we could rename it to `some` so that it follows the naming of the commonly used lodash utility, like some of the other helpers are already doing.

JemBijoux commented:
That would work for me. I know you also suggested maybe using `_.some` rather than the array filter in there. Not sure if it would warrant changing the body of this function though.

Another thought I had was a name like `ifAnyOf` (but that's probably becoming overly verbose).

JemBijoux commented:
The more I think about it, utilizing `_.some` probably does make more sense, so I'm happy to modify that. But that would assume we're okay with adding this helper, which it seems like there's still uncertainty there (and around what exactly to name it)?

On a related point, I thought about also submitting a similar request for an `#ifAllOf` helper that would work like an `#and` filter then too. Same thing there... could make use of `_.some` but test if any of the arguments returned false (showing the else content in that case). Currently we can do something similar by nesting if statements within one another, but when you're checking for more than just a few properties things start to look a little funny (and it's difficult to work out any sort of "else display X" condition).

hegrec commented:
In the case of potentially adding a helper like `all` i think it would be a good idea to make this change overload `any` to accept more than one argument, like you have at the start of your implementation, but not add a completely new helper. If there is only one value and it is an array or object, you can scan its contents and return true if any of them are truthy, if more than one argument is passed, do just a quick scan on them all, which `_.some`/`_.any` should do. That way we only have two helpers to handle the two cases of `#any` and `#all` rather than four.

I definitely can see the utility in having helpers like this for very specific scenarios that you want conditionals for in templates.

JemBijoux commented:
Cool, so rather than submit this `#or` helper, make changes to the any helper to potentially work with multiple arguments. That makes sense. I can make some adjustments on this branch and push one more commit?

Are you saying it might still make sense to have an `#all` helper separately? Doesn't appear to be anything like that in here yet, so for that one I would need to submit a fresh helper. I'm happy to work something up and submit a separate pr to cover that too if it sounds good to you folks.

hegrec commented:
That sounds good to me, please squash the commits down so only one gets merged though.

JemBijoux commented:
:+1: For sure! I'll touch base before this afternoon. (Would you prefer the #all helper on a new branch/separate commit or also included in all this?)

hegrec commented:
You can open that one separately.

JemBijoux commented:
Okay! Updated this so it's now using the "any" helper, simply extending it if there is no option hash passed through. Within the function I've switched over to using the lodash any helper also for consistency.

`{{#any value1 value2 ..etc}}
So as long as there are no `key=value` applied, the any check will be against all the available arguments.

`{{#any obj key=value}}`
Will check against the singular object like how this any helper originally functioned.

JemBijoux commented on helpers/any.js:
```diff
@@ -13,15 +13,47 @@ internals.implementation = function(handlebars) {
+        var args = [],
+            opts,
+            predicate,
+            any;
```
took the liberty of bringing the declarations up top here on this one too.

hegrec commented:
:+1: Seems good to me, @mcampa or @mickr can one of you guys concur before we merge this?

mcampa commented:
:+1: LGTM too

mcampa commented:
@bc-mkatz need to add this to the docs

mickr commented:
:+1:
